internal/adminapi/endpoints.go

```diff
@@ -0,0 +1,83 @@
+package adminapi
+
+import (
+	"context"
+	"fmt"
+
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetURLsForService performs an endpoint lookup, using provided kubeClient
+// to list provided Admin API Service EndpointSlices.
+func GetURLsForService(ctx context.Context, kubeClient client.Client, service types.NamespacedName) (sets.Set[string], error) {
+	const (
+		defaultEndpointSliceListPagingLimit = 100
+	)
+
+	// Get all the EndpointSlices assigned to the provided service.
+	labelReq, err := labels.NewRequirement("kubernetes.io/service-name", selection.Equals, []string{service.Name})
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		addresses     = sets.New[string]()
+		continueToken string
+		labelSelector = labels.NewSelector().Add(*labelReq)
+	)
+	for {
+		var endpointsList discoveryv1.EndpointSliceList
+		if err := kubeClient.List(ctx, &endpointsList, &client.ListOptions{
+			LabelSelector: labelSelector,
+			Namespace:     service.Namespace,
+			Continue:      continueToken,
+			Limit:         defaultEndpointSliceListPagingLimit,
+		}); err != nil {
+			return nil, err
+		}
+
+		for _, es := range endpointsList.Items {
+			addresses = addresses.Union(AddressesFromEndpointSlice(es))
+		}
+
+		if endpointsList.Continue == "" {
+			break
+		}
```

Comment:
@pmalek, did you intend to assign to the `continueToken` variable defined at line 30 above? As written, you always submit an empty continuation token, so you don't accommodate two or more pages of slice entries.

@czeslavo, what do you think?

Reply:
Please see #3815.

Reply:
Yeah, I agree it's a bug. 

For context why it's called with a client not using the controller's cache, I believe `GetURLsForService` is used before the controller manager gets created and that happens only once at startup time.

Reply:
Don't we need to be calling on this every time a watch fires on the `endpointslices` resource, in order to learn of the current—and presumably changed—set of proxy pods?

Reply:
It turns out that [we _do_ read from the cache when the watches fire later](https://sourcegraph.com/github.com/Kong/kubernetes-ingress-controller@1dca7d1/-/blob/internal/controllers/configuration/kongadminapi_controller.go?L98-L104), over in `(*internal/controllers/configuration.KongAdminAPIServiceReconciler).Reconcile`.

I was in the middle of filing a new issue proposing that we use watch and cache here, but it turns out that it's already in place.